### PR TITLE
Refactor HIP backend

### PR DESCRIPTION
**Pull request by Rombur**
This is a large PR but it is just moving code around. Everything that was not a memory space was moved out of `Kokkos_HIP_Space` and both `Kokkos_HIP.hpp` and `Kokkos_HIP_Space.hpp` have been move to `HIP` like we have done for the `OpenACC` backend.

### CODE REVIEW

File: core/src/decl/Kokkos_Declare_HIP.hpp [hunk 1]
```diff
@@ -46,7 +46,27 @@
...
+#include <HIP/Kokkos_HIP_UniqueToken.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+using HIPSpace           = ::Kokkos::HIPSpace;
+using HIPHostPinnedSpace = ::Kokkos::HIPHostPinnedSpace;
+using HIPManagedSpace    = ::Kokkos::HIPManagedSpace;
+using HIP                = ::Kokkos::HIP;
+}  // namespace Experimental
+}  // namespace Kokkos
```
Comment (dalg24) [inline core/src/decl/Kokkos_Declare_HIP.hpp:69]: Why aren't these in `<fwd/Kokkos_Fwd_HIP.hpp>`?

**Review by dalg24 — COMMENTED** [review]
Looks ok

File: core/src/decl/Kokkos_Declare_HIP.hpp (hunk 1 above)
Comment (Rombur) [reply core/src/decl/Kokkos_Declare_HIP.hpp:69]: What would I want to do that?

Comment (masterleinad) [reply core/src/decl/Kokkos_Declare_HIP.hpp:69]: Also, should we deprecate `Kokkos::Experimental::HIP*` to be removed in `Kokkos 5`?

Comment (Rombur) [reply core/src/decl/Kokkos_Declare_HIP.hpp:69]: > Also, should we deprecate Kokkos::Experimental::HIP* to be removed in Kokkos 5?

That's still open to discussion.

Comment (Rombur) [reply core/src/decl/Kokkos_Declare_HIP.hpp:69]: We could move `<fwd/Kokkos_Fwd_HIP.hpp>` but that means I would need to re-run the test suite with `Experimental::HIP` to ensure that `Experimental::HIP` still works.

Comment (dalg24) [reply core/src/decl/Kokkos_Declare_HIP.hpp:69]: OK you don't have to change

**Review by dalg24 — APPROVED** [review]

**Review by JBludau — APPROVED** [review]
looks good

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
What did you change?

**Comment by Rombur** [comment]
I had made a mistake when fixing a conflict with #5398 I just fixed it.

**Pull request merged**